lib/mobility/backends/active_record/container.rb

```diff
@@ -0,0 +1,116 @@
+require "mobility/backends/hash_valued"
+require "mobility/backends/active_record/pg_hash"
+
+module Mobility
+  module Backends
+=begin
+
+Implements the {Mobility::Backends::Container} backend for ActiveRecord models.
+
+=end
+    class ActiveRecord::Container
+      include ActiveRecord
+
+      require 'mobility/backends/active_record/container/query_methods'
+
+      # @return [Symbol] name of container column
+      attr_reader :column_name
+
+      # @!macro backend_constructor
+      # @option options [Symbol] column_name Name of container column
+      def initialize(model, attribute, options = {})
+        super
+        @column_name = options[:column_name]
+      end
+
+      # @!group Backend Accessors
+      #
+      # @note Translation may be a string, integer, boolean, hash or array
+      #   since value is stored on a JSON hash.
+      # @param [Symbol] locale Locale to read
+      # @param [Hash] options
+      # @return [String,Integer,Boolean] Value of translation
+      def read(locale, _ = nil)
+        model_translations(locale)[attribute]
+      end
+
+      # @!group Backend Accessors
+      # @note Translation may be string, integer or boolean-valued since
+      #   value is stored on a JSON hash.
+      # @param [Symbol] locale Locale to write
+      # @param [String,Integer,Boolean] value Value to write
+      # @param [Hash] options
+      # @return [String,Integer,Boolean] Updated value
+      def write(locale, value, _ = nil)
+        set_attribute_translation(locale, value)
+        model_translations(locale)[attribute]
+      end
+
+      def self.configure(options)
+        options[:column_name] ||= :translations
+      end
+
+      # @!macro backend_iterator
+      def each_locale
+        model[column_name].each do |l, _|
+          yield l.to_sym unless read(l).nil?
+        end
+      end
+
+      setup do |_attributes, options|
+        store options[:column_name], coder: Coder
+
+        # Fix for duping depth-2 jsonb column in AR < 5.0
+        if ::ActiveRecord::VERSION::STRING < '5.0'
+          column_name = options[:column_name]
+          module_name = "MobilityArContainer#{column_name.to_s.camelcase}"
+          unless const_defined?(module_name)
+            dupable = Module.new do
+              class_eval <<-EOM, __FILE__, __LINE__ + 1
+                def initialize_dup(source)
+                  super
+                  self.#{column_name} = source.#{column_name}.deep_dup
+                end
+              EOM
+            end
+            include const_set(module_name, dupable)
+          end
+        end
```

Comment:
This is a lot of code just to fix an issue which is arguably a bug in AR 4.2... debating maybe just skipping the dup specs for this backend in that version of Rails.